include/circt/Support/FVInt.h

```diff
@@ -0,0 +1,591 @@
+  /// this corresponds to SystemVerilog's `===` operator.
+  template <typename T>
+  bool operator==(const T &other) const {
+    return value == other && !hasUnknown();
+  }
+
+  /// Determine whether this integer is not equal to another.
+  bool operator!=(const FVInt &other) const { return !((*this) == other); }
+
+  /// Determine whether this integer is not equal to a two-valued integer.
+  template <typename T>
+  bool operator!=(const T &other) const {
+    return !((*this) == other);
+  }
+
+  //===--------------------------------------------------------------------===//
+  // String Conversion
+  //===--------------------------------------------------------------------===//
+
+  /// Convert a string into an `FVInt`.
+  ///
+  /// The radix can be 2, 8, 10, or 16. For radix 2, the input string may
+  /// contain the characters `x` or `X` to indicate an unknown X bit, and `z` or
+  /// `Z` to indicate an unknown Z bit. For radix 8, each X or Z counts as 3
+  /// bits. For radix 16, each X and Z counts as 4 bits. When radix is 10 the
+  /// input cannot contain any X or Z.
+  ///
+  /// Returns the parsed integer if the string is non-empty and a well-formed
+  /// number, otherwise returns none.
+  static std::optional<FVInt> tryFromString(StringRef str, unsigned radix = 10);
+
+  /// Convert a string into an `FVInt`. Same as `tryFromString`, but aborts if
+  /// the string is malformed.
+  static FVInt fromString(StringRef str, unsigned radix = 10) {
+    auto v = tryFromString(str, radix);
+    assert(v.has_value() && "string is not a well-formed FVInt");
+    return *v;
+  }
+
+  /// Convert an `FVInt` to a string.
+  ///
+  /// The radix can be 2, 8, 10, or 16. For radix 8 or 16, the integer can only
+  /// contain unknown bits in groups of 3 or 4, respectively, such that a `X` or
+  /// `Z` can be printed for the entire group of bits. For radix 10, the integer
+  /// cannot contain any unknown bits. In case the output contains letters,
+  /// `uppercase` specifies whether they are printed as uppercase letters.
+  ///
+  /// Appends the output characters to `str` and returns true if the integer
+  /// could be printed with the given configuration. Otherwise returns false and
+  /// leaves `str` in its original state. Always succeeds for radix 2.
+  bool tryToString(SmallVectorImpl<char> &str, unsigned radix = 10,
+                   bool uppercase = true) const;
+
+  /// Convert an `FVInt` to a string. Same as `tryToString`, but directly
+  /// returns the string and aborts if the conversion is unsuccessful.
+  SmallString<16> toString(unsigned radix = 10, bool uppercase = true) const {
+    SmallString<16> str;
+    bool success = tryToString(str, radix, uppercase);
+    assert(success && "radix cannot represent FVInt");
+    return str;
+  }
+
+  /// Print an `FVInt` to an output stream.
+  void print(raw_ostream &os) const;
+
+private:
+  APInt value;
+  APInt unknown;
```

Comment:
This is super exciting to see!! One nit: the Verilog standard document uses 'unknown' as the name specifically for an `x` bit:

> 0—represents a logic zero or a false condition
> 1—represents a logic one or a true condition
> x—represents an unknown logic value
> z—represents a high-impedance state

(from the [IEEE standard document](https://standards.ieee.org/ieee/1800/7743/))


 Is there a chance it could cause confusion that here 'unknown' refers to both `x` and `z` bits?

Reply:
That's a great point! I took the naming scheme from Slang's `SVInt`. This description of `x` as being _the_ unknown logic value is a bit unfortunate, because the SV standard in other places talks about functions like `$isunknown`, which checks for X _and_ Z. I guess that's also where Slang got its name from. My gut feeling would be to stick with Slang's naming scheme here.